### ISSUE TITLE
Averaging label representations

### ISSUE THREAD

**Issue by trpstra** [EMNLP2019LSAN/LSAN#5]
Hi,

Thanks for this work, it's an interesting idea.
I am a bit confused about the final averaging step in your forward method though, this line:

https://github.com/EMNLP2019LSAN/LSAN/blob/1d130c525861280167c99d8e11473222258ef709/attention/model.py#L84

Doesn't averaging all the label representations undo the effect of adaptive fusion? Your code comment says it's also possible to do a linear layer here instead of averaging. Can you please explain more? Just concat'ing the label representations would require the output layer to have shape (batch,  num_classes * emb_dim * 2, num_classes), which would never fit on any GPU's memory for any reasonable number of classes. Maybe some max pooling step would help here? Curious to hear your thoughts.
Cheers.

**Comment by AndrewLim1990** [comment]
Hey I just bumped into this question too. 

I don't think that averaging undoes any of the adaptive fusion calculation.

I think what's going on is that the `doc` is of shape: `(batch_size, num_classes, emb_dim)`. By taking `torch.sum(doc, 1)`, obtaining a singular document embedding by averaging across all of the "label specific" documentation embeddings.

This is different that what was mentioned in the paper. Not sure which version (average vs linear layer) was used to obtain the results mentioned in the paper. Has anyone been able to replicate the results?

**Comment by AndrewLim1990** [comment]
If anyone else was curious, the authors wrote a very similar paper with the corresponding code: https://github.com/HX-idiot/Hybrid_Attention_XML/blob/master/hybrid_xml.py

There, you can see the linear layer implemented
